### PR TITLE
Wallet Contract placeholder

### PR DESCRIPTION
**Pull request by staffik**
## Context
NEP: https://github.com/near/NEPs/issues/518
Tracking issue: https://github.com/near/nearcore/issues/10018.

### Goal
We want the NEAR Protocol's ecosystem to be closer with Ethereum by integrating Web3 wallet support.
To accomplish that, some changes on the protocol level are needed with an emphasis on user experience continuity.
Following the design, the main change on the protocol level would be to have ETH-style addresses managed by a new `Wallet Contract` integrated into the protocol. For seamless onboarding of EVM users, the contract will be free of charge.

### Previous work
This PR is built on top of two PRs:
* https://github.com/near/nearcore/pull/10020: no-op PR, laying groundwork for further changes.
* https://github.com/near/nearcore/pull/10224: empty contract as a placeholder, literally deployed to new ETH accounts.

## Summary
This PR adds `near-wallet-contract` crate (based on `runtime/near-test-contracts`) that exposes `Wallet Contract` WASM code to the runtime.
It stops deploying a copy of the `Wallet Contract` to each newly created ETH-implicit account.
Instead, it uses an in-memory cached contract code on `execute_function_call` action from such account.

### Changes
- Add `wallet-contract` crate (separated from the workspace) with placeholder `Wallet Contract` implementation.
- Add `near-wallet-contract` crate (part of the workspace) that generates (through `build.rs`) and exposes the `Wallet Contract` WASM code.
- Do not literally deploy `Wallet Contract` when creating ETH-implicit account. Just store reference to the `Wallet Contract`.
- Treat ETH-implicit accounts specially when retrieving contract code from an account (in such case returns in-memory cached `Wallet Contract` code).
- Add tests calling `Wallet Contract` where a transfer from an ETH-implicit account is possible depending on the public key passed with `rlp_transaction` argument.

### CODE REVIEW

**Comment by staffik** [comment]
I am not sure how `Wallet Contract` upgrade should be supported.
E.g. do we want to be able to dynamically switch the WASM file used, without recompiling?

File: runtime/near-wallet-contract/res/nightly_wallet_contract.wasm
Comment (birchmd) [inline runtime/near-wallet-contract/res/nightly_wallet_contract.wasm:1]: I don't think we should commit generated files to the repository. These files should always be created through the `build.rs` process, so they do not need to be committed. And committing them may cause issues if the Wallet Contract code is changed without the corresponding updated to these binaries.

File: runtime/near-wallet-contract/src/lib.rs
Comment (birchmd) [inline runtime/near-wallet-contract/src/lib.rs:15]: In the implementation of `get_contract_code` this nightly variant is never used. Maybe instead of having two version of the `wallet_contract` function, there should only be one definition and the implementation will depend on the `nightly` compile-time feature flag.

File: runtime/runtime/src/actions.rs
Comment (birchmd) [inline runtime/runtime/src/actions.rs:491]: It was briefly discussed on Zulip that we could have a placeholder value (as opposed to real Wasm code) that is still set. The reason to do this would be to still signal to tooling like explorers that there is code deployed to these accounts (as opposed to accounts where there is no code at all), it's just that the deployed code is not explicitly specified in the state. It also could be a baby step towards a future generalized code sharing feature.

That said, I like the simplicity of simply storing nothing and using the convention that all Eth implicit accounts have an implicit contract also. Mostly this comment is here to kick off a discussion with the Runtime team about the design choice. Personally I am in favor of the option you have chosen here.

File: integration-tests/src/tests/client/features/delegate_action.rs
Comment (birchmd) [inline integration-tests/src/tests/client/features/delegate_action.rs:961]: Do I understand correctly that the reason to use meta-transactions here is just because of the nature of the placeholder contract not having its own signature checking? So in the future when we have the real Wallet Contract we will be able to remove this test an replace it with tests that directly call the Wallet Contract?

The reason I as is because in the design of the overall solution I do not think we will be using Near meta-transactions. So I want to make sure there is no assumption in the code that meta-transactions are important to the transaction flow.

**Comment by birchmd** [comment]
> I am not sure how Wallet Contract upgrade should be supported. E.g. do we want to be able to dynamically switch the WASM file used?

I do not think we want to support dynamic upgrades. It would be a complex feature to get right because you still need to have all the nodes agree on what the right Wallet Contract binary to use is, so you would need its hash to be a property of the epoch or something like that.

Upgrades will always have to be protocol upgrades. Both the old and new binaries will need to be kept and the node would need to switch which one is used depending on the protocol version.

**Comment by wacban** [comment]
Sorry for the delay, I was busy with investigations, I'll have a look on Monday.

File: integration-tests/src/tests/client/features/delegate_action.rs
Comment (staffik) [reply integration-tests/src/tests/client/features/delegate_action.rs:961]: Yes, exactly. I've just added a comment saying that this test should be replaced once Wallet Contract is ready.
There is no assumption about meta-transactions in the code.

File: runtime/runtime/src/actions.rs
Comment (staffik) [reply runtime/runtime/src/actions.rs:491]: Do you remember which topic the discussion happened?
I thought it is https://near.zulipchat.com/#narrow/stream/295306-pagoda.2Fcontract-runtime/topic/Eth-wallet.20support but I do not see it there.

Comment (birchmd) [reply runtime/runtime/src/actions.rs:491]: Yes, that's the one I was referring to. Though, again, I am happy with what is already implemented in this PR. It's only a question of if someone from Runtime teams thinks it is important to push for having some kind of code represented in the state.

File: runtime/near-wallet-contract/src/lib.rs
Comment (wacban) [reply runtime/near-wallet-contract/src/lib.rs:14]: What I would reccomed us to do is to discuss how upgrades will work and make sure that it's going to be possible to add this support painlessly in the future but without implemeting full support for it from day one. 

If we need to add some sort of versioning let's do it from day one. It could be some property of the blockchain at the time the eth implicit account was created, stored in the account itself. It can be the hash of the contract, the protocol version, some new dedicated version just for the wallet contract or something else.

File: runtime/near-wallet-contract/wallet-contract/src/lib.rs
Comment (wacban) [inline runtime/near-wallet-contract/wallet-contract/src/lib.rs:19]: ❤️

File: runtime/near-wallet-contract/wallet-contract/src/lib.rs
Comment (wacban) [inline runtime/near-wallet-contract/wallet-contract/src/lib.rs:23]: Is there some sort of schema library that we could use to automatically deserialize the fields for us? 

Doesn't need to be in this PR, just generally speaking it would be nice to separate the deserializing from the business logic validation and then the execution.

File: runtime/runtime/src/actions.rs
Comment (wacban) [inline runtime/runtime/src/actions.rs:47]: ❤️

Comment (wacban) [reply runtime/runtime/src/actions.rs:47]: How does charging gas work here and how should it work here?

File: runtime/runtime/src/actions.rs
Comment (wacban) [inline runtime/runtime/src/actions.rs:507]: Nice! This should allow us to upgrade the contracts in the future as needed.

File: runtime/runtime/src/actions.rs
Comment (wacban) [inline runtime/runtime/src/actions.rs:495]: Sanity check - why do we not need precompilation?

File: runtime/runtime/src/actions.rs
Comment (staffik) [reply runtime/runtime/src/actions.rs:491]: @Ekleog Could you take a look at this PR as someone from the Runtime team?

Comment (Ekleog-NEAR) [reply runtime/runtime/src/actions.rs:491]: I submitted this as a comment on near/NEPs#518 :)

And this brings me to this PR. Should we actually review it before near/NEPs#518 lands? This feels to me like doing the mistake we did near/NEPs#483 again, and we should not land any implementation before agreeing on the design.

(Also, I’d suggest adding near/NEPs#518 to the top post, as AFAIU this PR is supposed to be implementing it.)

**Comment by Ekleog-NEAR** [comment]
As a first operational comment, I’ll say that compiling wasm in the neard build process can be problematic for some. In particular, https://github.com/NixOS/nixpkgs/commit/ac9b3799d2f7b43cb59072f87594a322b26c195d was done outside of us to remove near-test-contracts from the compilation stage.

Can we check in the compiled wasm code alongside near-wallet-contract, and instead of recompiling on each build, just checking that the compiled file is the same as the provided file on each build? This way, downstream users who don’t want to deal with a wasm32 toolchain can just disable that check.

Now, it’s most certainly not a big deal if we can’t do it, considering that all production validators would definitely want to recompile the wasm blob anyway.

File: runtime/near-wallet-contract/README.md
Comment (staffik) [reply runtime/near-wallet-contract/README.md:8]: Yes, it is not included in the repository now and CI seems to include it since the test passes.

File: runtime/near-wallet-contract/build.rs
Comment (staffik) [reply runtime/near-wallet-contract/build.rs:51]: I am also surprised :)

File: runtime/near-wallet-contract/src/lib.rs
Comment (staffik) [reply runtime/near-wallet-contract/src/lib.rs:14]: > Just for my full understanding, why do you need support for nightly here?
> 
> I'm fully onboard being future proof and supporting upgrading the contract but unless you need it now I would keep it simple and skip that part. I'm happy to be convinced otherwise though :)

I removed support for nightly. It not used currently and it is easy to add.



> What I would reccomed us to do is to discuss how upgrades will work and make sure that it's going to be possible to add this support painlessly in the future but without implemeting full support for it from day one.
> 
> If we need to add some sort of versioning let's do it from day one. It could be some property of the blockchain at the time the eth implicit account was created, stored in the account itself. It can be the hash of the contract, the protocol version, some new dedicated version just for the wallet contract or something else.

Will discuss it at the today's meeting.

File: runtime/near-wallet-contract/wallet-contract/src/lib.rs
Comment (Ekleog-NEAR) [inline runtime/near-wallet-contract/wallet-contract/src/lib.rs:1]: The fact that this implementation passes the integration test above makes me think: when making this production-ready, we should also have at least a test that has the right public key but an invalid signature

File: runtime/runtime/src/actions.rs
Comment (Ekleog-NEAR) [inline runtime/runtime/src/actions.rs:59]: I guess this is OK for now, as it’s unstable and the discussion on NEP-518 has not stabilized yet, but I do think that we’ll want something more stable, for users, than just "literally any eth implicit account lives with the latest version of the eth contract"

Also, you seem to have been careful to have an eth account be zero-balance-ready. Is there any use case for it? If yes, we should probably add that as a test, as currently the only test does have 1 yoctoNEAR and thus was already created before entering this function.

Disclaimer: I’m saying this because I’m not sure how the code would behave if you tried to call a function on the wallet contract without having funded the account first

**Review by Ekleog-NEAR — CHANGES_REQUESTED** [review]
It’s taken me a while, but here is my review on this code! Thank you for it :)

The solution of pulling the crate out of the workspace works fine for near-test-contracts, which we do not really care about. However, for wallet-contract, we definitely do care a lot about it being correct. Have you tried having it within the workspace, which would allow all our CI tooling to also apply to it, and thus all the related linters?

I think that, before stabilizing anything here, we must make sure that:
1. near/NEPs#518 has landed
2. We have a clear idea of how upgrades to the wallet contract should happen. In particular, this means, do we want to keep old wallet contract versions around, like we would for other global contract code? But this discussion will probably be better on near/NEPs#518 directly

Marking this PR as needing changes, for at least:
- the zeroize downgrade
- `Clone` on `ContractCode`
- the extraneous dependencies
- `near-wallet-contract` not working without a full build tree
- overflow-checks / debug-assertions, as we’ll forget to come back to then with future changes

The rest are improvements that I think are worth it, but probably do not deserve blocking this PR :)

File: integration-tests/src/tests/client/features/delegate_action.rs
Comment (birchmd) [reply integration-tests/src/tests/client/features/delegate_action.rs:991]: I do not think such a PR will be necessary. When the real version of the Wallet Contract is in place the integration tests will not be trying to encode `u128` values directly (any Near balances will be upgraded to 256-bit values because that is the Ethereum standard).

File: runtime/runtime/src/actions.rs
Comment (birchmd) [reply runtime/runtime/src/actions.rs:59]: I think we should assume that Eth implicit accounts be have the same a the current Near implicit accounts. I think the behaviour for these is to reject `FunctionCall` actions if the account does not exist, but `Transfer` actions can create the account. `Transfer { deposit: 0 }` is allowed with zero-balance accounts so you can create implicit accounts for free.

File: runtime/near-wallet-contract/wallet-contract/Cargo.toml
Comment (staffik) [reply runtime/near-wallet-contract/wallet-contract/Cargo.toml:25]: Are not `debug-assertions` meant to only be available in debug mode? These are settings for `[profile.release]` so I think we should disable `debug-assertions` there.

File: runtime/near-wallet-contract/wallet-contract/src/lib.rs
Comment (staffik) [reply runtime/near-wallet-contract/wallet-contract/src/lib.rs:1]: Yes, we definitely must have. A plenty of new tests will be added when the real Wallet Contract implementation is ready. And there will be audit process.

File: runtime/runtime/src/actions.rs
Comment (staffik) [reply runtime/runtime/src/actions.rs]: `meta_tx_create_implicit_account` test shows that we can create eth-implicit account with  `Transfer { deposit: 0 }`.
I've just added a test (`meta_tx_call_wallet_contract_account_does_not_exist`) that shows that `FunctionCall` action is rejected if the account does not exist.

File: integration-tests/src/tests/client/features/delegate_action.rs
Comment (staffik) [reply integration-tests/src/tests/client/features/delegate_action.rs]: I've added a comment so that we do not forget about it.

File: runtime/runtime/src/actions.rs
Comment (staffik) [reply runtime/runtime/src/actions.rs:491]: Following @Ekleog-NEAR suggestion [here](https://near.zulipchat.com/#narrow/stream/295306-pagoda.2Fcontract-runtime/topic/Eth-wallet.20support/near/403646004) I've changed it to deploy near[wallet contract hash]. ETH-implicit accounts will now store hash of these magic bytes.

File: Cargo.lock
Comment (staffik) [reply Cargo.lock]: I have no idea. Seems like without a reason. I reverted it back manually to 1.6.0 and checked if it builds without warnings.

**Comment by staffik** [comment]
Here are the all [review changes](https://github.com/near/nearcore/compare/b138347..07a9069), mainly:
- https://github.com/near/nearcore/pull/10269#pullrequestreview-1779536724
- Moved `Wallet Contract` tests to a separate file and added some tests.
- `near[wallet contract hash]` is deployed as a contract code to new ETH-implicit accounts.
- `Wallet Contract` is checked into the repository and it is included into the binary at compile time. By default it is not recompiled, one can recompile it with `cargo build --features build_wallet_contract`.

> The solution of pulling the crate out of the workspace works fine for near-test-contracts, which we do not really care about. However, for wallet-contract, we definitely do care a lot about it being correct. Have you tried having it within the workspace, which would allow all our CI tooling to also apply to it, and thus all the related linters?

@Ekleog-NEAR I've built this PR with the `wallet-contract` crate being part of the workspace, and I liked the linter.
But after the PR was ready I removed the crate with [this commit](https://github.com/near/nearcore/pull/10269/commits/16143464ed36168d6c15f178b2d30778e86ca775) because crazy things happened with Cargo.lock.
The crate is only used during the build (now it is not even used during build by default) and it introduces `near-sdk` dependency to the workspace which makes all that crazy stuff happen with the `Cargo.lock`. I spent some time trying to fix that but without success.

**Review by Ekleog-NEAR — APPROVED** [review]
For the vulnerable dependencies, that’s something we definitely need to fix on the near-sdk side, because right now we’re pushing everyone to use these vulnerable dependencies.

For wee-alloc, the change recommended by [the rustsec advisory](https://rustsec.org/advisories/RUSTSEC-2022-0054) is to just use Rust’s default allocator. For ed25519, bumping near-crypto & co to 0.18 (as soon as I can get it released, currently we’re facing a few issues with the release) should deal with the issue. I’d suggest fixing the wee-alloc issue, and ignoring the red cargo-audit checkmark for ed25519 for now, considering it’s dependencies that only get used for non-wasm32 targets.

Anyway, I think we can merge this as is for now, and merge the wallet contract into the workspace as a follow-up PR :)

I think the only comment here that really needs action before landing is, removing the prebuilt .wasm file, because it’s currently impossible to reproduce. It’ll make operational deployment harder, but if we check in a binary file, we definitely need it to be reproducible.

In order to keep operational deployment easy for now, I’d suggest compiling the include_bytes! conditionally on the nightly feature flag only (and eg. having a hardcoded empty slice on stable), this way we can gain some time to decide what the proper process should be there :)

File: runtime/near-wallet-contract/README.md
Comment (staffik) [reply runtime/near-wallet-contract/README.md]: @Ekleog-NEAR
I've just removed the precompiled contract and I included empty WASM file (to make `include_bytes` happy). It uses hardcoded empty slice on stable and it rebuilds the contract on nightly.

Comment (Ekleog-NEAR) [reply runtime/near-wallet-contract/README.md]: Sounds good to me! I’m surprised that you needed an empty WASM file on nightly, considering that `build.rs` should be building it anyway on nightly, but I guess some tooling like editors might miss that subtlety :)

File: runtime/near-wallet-contract/build.rs
Comment (birchmd) [inline runtime/near-wallet-contract/build.rs:42]: The way this contract is built will need to be changed for it to have a reproducible hash. By default Rust compilation includes things like the absolute paths for the source code compiled, which obviously differ between machines. Near publishes a [Docker container](https://hub.docker.com/r/nearprotocol/contract-builder) with a standardized build environment to achieve reproducible builds.

We can handle this in a follow-up PR though I suppose.

File: runtime/near-wallet-contract/src/lib.rs
Comment (birchmd) [inline runtime/near-wallet-contract/src/lib.rs:45]: Should these hash-checking tests be marked with `ignore` until the contract can be built with a reproducible hash?

**Review by birchmd — APPROVED** [review]
Looks good to me! We only need to address the issue of reproducible build for the contract, and I think that can be done as a follow-up PR.

File: runtime/near-wallet-contract/build.rs
Comment (nagisa) [reply runtime/near-wallet-contract/build.rs:44]: There’s the `--remap-path-prefix` option for rustc that should enable reproducible builds.

Comment (staffik) [reply runtime/near-wallet-contract/build.rs:44]: Thank you @nagisa! I was not aware of that.

**Pull request merged**
